python_modules/dagster/dagster/_core/definitions/unresolved_asset_job_definition.py

```diff
@@ -255,7 +266,11 @@ def define_asset_job(
             location. A list of strings represents the union of all assets selected by strings
             within the list.
 
-            The selection will be resolved to a set of assets once the when location is loaded.
+            The selection will be resolved to a set of assets when the location is loaded. If the
```

Comment:
Two things:

1. Running source asset observations together in the same job with materializations doesn't currently work. It could be made to work with a little more effort, but...
2. If we allow this, people are going to expect the source asset observations to affect downstream materializations in the same run. Therefore I don't think we should enable this until that functionality is in place, which is a separate problem ("runtime staleness"). Keeping it this way guides users to run their source asset observations _before_ materializing their assets.

Reply:
I see. So is the idea that, once we add support for running source asset observations in the same job with materializations and we enable skipping steps mid-run based on memoization, we then _would_ make `AssetSelection.all()` observe all the source assets and materialize all the regular assets?

Reply:
yes

Reply:
That makes sense.

Once we make that change, the behavior of `AssetSelection.all()` would change from the current behavior, right?

If that's the case, it might be better to raise an error now in the case that a selection includes both source assets and regular assets, to avoid a behavior change that users might miss.

Reply:
Agreed-- made that change.

Reply:
> So, I experimented with doing this (you can see it in the current state of the branch), but this may break existing users because AssetSelection.upstream() currently includes source assets.

This would only break existing users with observable source assets, right? I think we're technically allowed to make this change, because observable source assets are experimental.

Reply:
> AssetSelections were not supposed to be able to include source assets but they can via upstream()

I think we should consider this a bug. Not just suggesting this because it's convenient here - I think it's actually unexpected behavior in the current system.

Reply:
OK-- I moved the adjustments to `AssetSelection` into a separate PR downstack that targets this bugfix among other things: https://github.com/dagster-io/dagster/pull/13027